datafusion/expr/src/type_coercion/functions.rs

```
@@ -136,62 +137,100 @@ fn maybe_data_types(
+            Some(type_into)
+        }
+        Float32
+            if matches!(
                 type_from,
-                Null | Timestamp(_, _) | Date32 | Utf8 | LargeUtf8
-            )
+                Null | Int8
+                    | Int16
+                    | Int32
+                    | Int64
+                    | UInt8
+                    | UInt16
+                    | UInt32
+                    | UInt64
+                    | Float32
+            ) =>
+        {
+            Some(type_into)
         }
-        Interval(_) => {
-            matches!(type_from, Utf8 | LargeUtf8)
+        Float64
+            if matches!(
+                type_from,
+                Null | Int8
+                    | Int16
+                    | Int32
+                    | Int64
+                    | UInt8
+                    | UInt16
+                    | UInt32
+                    | UInt64
+                    | Float32
+                    | Float64
+                    | Decimal128(_, _)
+            ) =>
+        {
+            Some(type_into)
+        }
+        Timestamp(TimeUnit::Nanosecond, None)
+            if matches!(
+                type_from,
+                Null | Timestamp(_, None) | Date32 | Utf8 | LargeUtf8
+            ) =>
+        {
+            Some(type_into)
         }
-        Utf8 | LargeUtf8 => true,
-        Null => can_cast_types(type_from, type_into),
-        _ => false,
+        Interval(_) if matches!(type_from, Utf8 | LargeUtf8) => Some(type_into),
+        Utf8 | LargeUtf8 => Some(type_into),
+        Null if can_cast_types(type_from, type_into) => Some(type_into),
+
+        // timestamp coercions, with timezone, accept the type_from timezone if valid
+        Timestamp(TimeUnit::Nanosecond, Some(_))
+            if matches!(
+                type_from,
+                Timestamp(TimeUnit::Nanosecond, Some(from_tz)) if arrow_array::timezone::Tz::from_str(from_tz).is_ok()
```

This is why/where added the arrow_array dependency. Could have done it differently, but this is the exact method/src used later on -- so hoping to fails here during type coercion. Maybe that's not the right way? 🤷🏼

---

Also, this may not be "the fix" long term -- as it's been [mentioned elsewhere](https://github.com/apache/arrow-datafusion/pull/6818#discussion_r1251657851) that the coercion needs refactoring. Can this work as an interim approach?

---

I think using the `arrow_array` function to parse strings to a valid timezone is the correct choice

Also I don't think `arrow_array` is a new dependency as  `datafusion_expr` already depends on the `arrow` crate, which depends on `arrow_array`. Thus I think adding a new explicit dependency to use `arrow_array::timezone::Tz` is fine

---

It would be nice if we didn't have to discard the `ArrowError` (via `is_ok`) as that causes the work to create the error message to bis discarded.

However, I think this will need a new API in arrow-rs -- so I'll file a new ticket there prior to merging this PR

---

Now that I think about this I wonder if there is any reason to check for valid timezones here at all -- if there is an invalid timezone, any actual calculation will fail susbequently

Perhaps we can sidestep the whole "don't ignore error" thing entirely if we just skipped the check 🤔

---

Ah, yes! We would have had a parser error if the tz string was invalid. Excellent point.